### PR TITLE
Creation of shadow-ui root element is fixed (closes #934)

### LINKED ISSUE #934: Shadow-ui root element isn't the last one
Page: http://todomvc.com/examples/dojo/
Related with https://github.com/DevExpress/testcafe/issues/958

### PR DESCRIPTION
**Pull request by helen-dikareva**
/cc @churkin @AlexanderMoskovkin @MarinaRukavitsyna

### CODE REVIEW

File: src/client/sandbox/node/index.js [hunk 1]
```diff
@@ -67,7 +67,8 @@ export default class NodeSandbox extends SandboxBase {
     // resources. Our goal is to make the native script think that all resources are fetched from the destination
     // resource, not from proxy, and also provide proxying for dynamically created elements.
     attach (window) {
-        var document = window.document;
+        var document                    = window.document;
+        var DOMContentLoadedEventRaised = false;
```
Comment (AlexanderMoskovkin) [inline src/client/sandbox/node/index.js]: domContentLoadedEventRaised

File: src/client/sandbox/node/index.js [hunk 2]
```diff
@@ -85,14 +86,28 @@ export default class NodeSandbox extends SandboxBase {
         // NOTE: In Google Chrome, iframes whose src contains html code raise the 'load' event twice.
         // So, we need to define code instrumentation functions as 'configurable' so that they can be redefined.
         Object.defineProperty(window, INTERNAL_PROPS.processDomMethodName, {
-            value:        (el, doc) => this.processNodes(el, doc),
+            value: (el, doc) => {
+                // NOTE: testcafe creates shadow-ui root before DOMContentLoaded event (once document.body is available).
+                // Sometimes for very heavy DOM or very slow loading at that moment body doesn't contain all elements
```
Comment (MarinaRukavitsyna) [inline src/client/sandbox/node/index.js]: // Sometimes for a very heavy DOM or a very slow loading at that moment, the body doesn't contain all elements

File: src/client/sandbox/node/index.js [hunk 3]
```diff
@@ -85,14 +86,28 @@ export default class NodeSandbox extends SandboxBase {
...
+                // and as a result after full page loading our root element become not the last child of body. So we need
```
Comment (MarinaRukavitsyna) [inline src/client/sandbox/node/index.js]: // and as a result after a full page loading our root element becomes not the last child of the body. So we need

File: src/client/sandbox/node/index.js [hunk 4]
```diff
@@ -85,14 +86,28 @@ export default class NodeSandbox extends SandboxBase {
...
+                // to make root last body child manually on every script loading until DOMContentLoaded event raise.
```
Comment (MarinaRukavitsyna) [inline src/client/sandbox/node/index.js]: // to make the root last body child manually on every script loading until the DOMContentLoaded event is raised.

File: src/client/sandbox/shadow-ui.js [hunk 5]
```diff
@@ -254,29 +257,33 @@ export default class ShadowUI extends SandboxBase {
...
-                this.onBodyContentChanged();
+                this.makeRootLastBodyChild();
         });
     }
 
-    onBodyContentChanged () {
-        if (this.root) {
-            if (!domUtils.closest(this.root, 'html'))
-                this.nativeMethods.appendChild.call(this.document.body, this.root);
+    makeRootLastBodyChild () {
```
Comment (AlexanderMoskovkin) [inline src/client/sandbox/shadow-ui.js]: I suppose it should be `onBodyElementMutation` (discussed in private with @helen-dikareva )

File: src/client/sandbox/shadow-ui.js [hunk 6]
```diff
@@ -254,36 +257,33 @@ export default class ShadowUI extends SandboxBase {
...
-    onBodyElementMutation () {
-        if (this.root) {
-            if (this.document.body && this.root.parentNode !== this.document.body)
-                this.nativeMethods.appendChild.call(this.document.body, this.root);
+        if (this.root.nextElementSibling) {
+            this.nativeMethods.appendChild.call(this.document.body, this.root);
+            return;
         }
+
+        // NOTE: Fix for B239138 - unroll.me 'Cannot read property 'document' of null' error raised during recording
```
Comment (MarinaRukavitsyna) [inline src/client/sandbox/shadow-ui.js]: // NOTE: Fix for B239138 - The 'Cannot read property 'document' of null' error is thrown on recording on the unroll.me site
// There was an issue when document.body was replaced, so we need to reattach a UI to a new body manually.

**Comment by helen-dikareva** [comment]
@testcafe-build-bot \retest

**Comment by helen-dikareva** [comment]
@testcafe-build-bot \retest

**Comment by helen-dikareva** [comment]
FPR

**Pull request merged**